org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPPresentationReconcilingStrategy.java

```diff
@@ -0,0 +1,316 @@
+			if (viewer == null)
+				return;
+
+			// fireInstall(viewer, newDoc);
+			try {
+				viewer.addTextListener(this);
+				// Update the grammar
+				// final IGrammar localGrammar = findGrammar(newDoc);
+				//
+				// if (localGrammar != null) {
+				// LSPPresentationReconciler.this.grammar = localGrammar;
+				// } else if (isThrowError()) {
+				// throw new TMException("Cannot find TextMate grammar for the given document");
+				// }
+
+				// Update the theme
+				// if (localGrammar != null) {
+				// final String scopeName = localGrammar.getScopeName();
+				// if (tokenProvider == null) {
+				// tokenProvider = TMUIPlugin.getThemeManager().getThemeForScope(scopeName,
+				// viewer.getTextWidget().getBackground().getRGB());
+				// }
+				// if (tokenProvider != null) {
+				// applyThemeEditor();
+				// } else if (isThrowError()) {
+				// throw new TMException("Cannot find Theme for the given grammar '" + scopeName
+				// + "'");
+				// }
+				// }
+				//
+				// if (localGrammar != null) {
+				// final var enable = LSPPresentationReconciler.this.enabled = tokenProvider !=
+				// null;
+				// if (enable) {
+				// // Connect a TextModel to the new document.
+				// final var docModel = TMModelManager.INSTANCE.connect(newDoc);
+				// docModel.setGrammar(localGrammar);
+				//
+				// // Add model listener
+				// docModel.addModelTokensChangedListener(this);
+				// }
+				// } else {
+				// LSPPresentationReconciler.this.enabled = false;
+				// }
+			} catch (final Exception e) {
+				// Platform.getLog(Platform.getBundle(TMEclipseRegistryPlugin.PLUGIN_ID)).log(
+				// new Status(IStatus.ERROR, TMUIPlugin.PLUGIN_ID, "Error while initializing
+				// TextMate model.", e));
+			}
+		}
+
+		@Override
+		public void textChanged(final @Nullable TextEvent event) {
+			if (event == null || !event.getViewerRedrawState()) {
+				return;
+			}
+
+			final var viewer = LSPPresentationReconcilingStrategy.this.viewer;
+			if (viewer == null)
+				return;
+
+			final IDocument doc = viewer.getDocument();
+			if (doc == null) {
+				return;
+			}
+
+			final IDocument document = viewer.getDocument();
+			URI uri = LSPEclipseUtils.toUri(document);
+			if (uri == null) {
+				return;
+			}
+			final DocumentLinkParams params = new DocumentLinkParams(new TextDocumentIdentifier(uri.toString()));
+			LanguageServiceAccessor
+					.getLanguageServers(document, capabilities -> capabilities.getDocumentLinkProvider() != null)
+					.thenApplyAsync(languageServers -> {
+						List<CompletableFuture<List<DocumentLink>>> futures = new ArrayList<>();
+						for (LanguageServer languageServer : languageServers) {
+							futures.add(languageServer.getTextDocumentService().documentLink(params));
+						}
+						allOf(futures).thenAccept(links -> {
+							final Control control = viewer.getTextWidget();
+							if (control != null) {
+								control.getDisplay().asyncExec(() -> colorize(links));

```

Comment:
I believe all the computation can be run in any thread, and only the `viewer.changeTextPresentation()` should be in the UI Thread.

Reply:
I need to do that otherwise I have an invalid SWT Thread access.